src/peerbook/libp2p_peerbook.erl

```diff
@@ -77,22 +77,40 @@ put(#peerbook{tid=TID, stale_time=StaleTime}=Handle, PeerList) ->
     ThisPeerId = libp2p_swarm:pubkey_bin(TID),
     %% XXX uncomment this to reject any peers publishing RFC1918 addresses once the network has transitioned over
     AllowRFC1918 = true, %% is_rfc1918_allowed(TID),
-    NewPeers = lists:filter(fun(NewPeer) ->
-                                    NewPeerId = libp2p_peer:pubkey_bin(NewPeer),
-                                    case unsafe_fetch_peer(NewPeerId, Handle) of
-                                        {error, not_found} -> AllowRFC1918 orelse not libp2p_peer:has_private_ip(NewPeer);
-                                        {ok, ExistingPeer} ->
-                                            %% Only store peers that are not _this_ peer,
-                                            %% are newer than what we have,
-                                            %% are not stale themselves
-                                            NewPeerId /= ThisPeerId
-                                                andalso (AllowRFC1918 orelse not libp2p_peer:has_private_ip(NewPeer))
-                                                andalso libp2p_peer:supersedes(NewPeer, ExistingPeer)
-                                                andalso not libp2p_peer:is_stale(NewPeer, StaleTime)
-                                                andalso not libp2p_peer:is_similar(NewPeer, ExistingPeer)
-                                                andalso libp2p_peer:network_id_allowable(NewPeer, libp2p_swarm:network_id(TID))
-                                    end
-                            end, PeerList),
+    NewPeers = lists:foldl(
+                 fun(NewPeer, Acc) ->
+                         NewPeerId = libp2p_peer:pubkey_bin(NewPeer),
+                         case unsafe_fetch_peer(NewPeerId, Handle) of
+                             {error, not_found} ->
+                                 case AllowRFC1918 orelse not libp2p_peer:has_private_ip(NewPeer) of
+                                     true -> [NewPeer | Acc];
+                                     false -> Acc
+                                 end;
+                             {ok, ExistingPeer} ->
+                                 %% Only store peers that are not _this_ peer,
+                                 %% are newer than what we have,
+                                 %% are not stale themselves
+                                 case NewPeerId /= ThisPeerId
+                                     andalso (AllowRFC1918 orelse not libp2p_peer:has_private_ip(NewPeer))
+                                     andalso libp2p_peer:supersedes(NewPeer, ExistingPeer)
+                                     andalso not libp2p_peer:is_stale(NewPeer, StaleTime)
+                                     andalso libp2p_peer:network_id_allowable(NewPeer, libp2p_swarm:network_id(TID))
+                                     andalso not libp2p_peer:is_similar(NewPeer, ExistingPeer) of
```

Comment:
Why not just remove the actual is_similar negative check ? If this is needed we need to refactor this a bit

Reply:
Yeah, I think this was left in as an oversight, will remove it.